### ISSUE TITLE
Backward compute time in DDP logging may not be accurate 

### ISSUE THREAD

**Issue by rohan-varma** [pytorch/pytorch#53427]
## 🐛 Bug

We start recording backwards compute time in `Reducer::prepare_for_backward`: https://github.com/pytorch/pytorch/blob/master/torch/lib/c10d/reducer.cpp#L901. However, prepare_for_backward is actually called at the end of the forward pass, and thus the model still generally has at least loss computation and could potentially do other operations before actually calling backwards. Hence the absolute time logged here may not be accurate. 

On the other hand, at least in cases where loss computation is relatively quick and we immediately go into bwd after forward pass, this inaccuracy may not be too much. But the case I'm concerned about is the following:
```
outputs = ddp(inputs) # runs reducer::prepare_for_bwd
# ... any arbitrary, long-running code
loss = CrossEntropy(outputs)
# possibly compute multiple losses
# ... any arbitrary, long-running code
loss.backward() # backwards actually starts 
```

**Comment by rohan-varma** [comment]
Discussed offline with a few folks and we can probably address this by adding a custom autograd function that executes before the regular backwards pass.
